File: cpp/1458.max-dot-product-of-two-subsequences.cpp
/*
 * @lc app=leetcode id=1458 lang=cpp
 *
 * [1458] Max Dot Product of Two Subsequences
 */
#include <vector>
#include <cstring>
using namespace std;
// @lc code=start
// t:O(n1*n2), s:O(n1*n2)
class Solution
{
public:
    int memo[501][501];
    int maxDotProduct(vector<int> &nums1, vector<int> &nums2)
    {
        memset(memo, -10000000, sizeof(memo));
        int n1 = nums1.size(), n2 = nums2.size();
        for (int i = 0; i != n1; ++i)
        {
            for (int j = 0; j != n2; ++j)
            {
                memo[i + 1][j + 1] = max({nums1[i] * nums2[j], memo[i + 1][j], memo[i][j + 1], memo[i][j] + nums1[i] * nums2[j]});
            }
        }
        return memo[n1][n2];
    }
};
// @lc code=end

File: py/1458.max-dot-product-of-two-subsequences.py
#
# @lc app=leetcode id=1458 lang=python3
#
# [1458] Max Dot Product of Two Subsequences
#
from typing import *
# @lc code=start
# t:O(m*n), s:O(m*n)

class Solution:
    def maxDotProduct(self, nums1: List[int], nums2: List[int]) -> int:
        bound = ~(1 << 31)
        n1 = len(nums1)
        n2 = len(nums2)
        dp = [[bound]*(n2+1) for _ in range(n1+1)]
        for i in range(1, n1+1):
            for j in range(1, n2+1):
                dp[i][j] = max(nums1[i-1]*nums2[j-1], dp[i-1][j-1]+nums1[i-1]
                               * nums2[j-1], dp[i-1][j], dp[i][j-1])
        return dp[n1][n2]
# @lc code=end
